src/charm.py

```diff
@@ -55,6 +58,13 @@ def __init__(self, *args):
 
         # Relation handler objects
 
+        # Self-monitoring
+        self._scraping = MetricsEndpointProvider(
+            self,
+            relation_name="self-metrics-endpoint",
+            jobs=[{"static_configs": [{"targets": [f"*:{self._port}"]}]}],
```

Comment:
I forget - where is the wildcard expanded? I.e. how does a remote scraper know the addresses of all the units?

Reply:
This happens [here](https://github.com/canonical/prometheus-k8s-operator/blob/a0efc401a2fbf65583cf6326e745d89a3c7efcbd/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1128). If it's a wildcard, it's passed on as a "labeled unit host" (which isn't really units anymore, but still). 

That method gets a list of hosts [from `_relation_hosts`](https://github.com/canonical/prometheus-k8s-operator/blob/a0efc401a2fbf65583cf6326e745d89a3c7efcbd/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1052), which pulls `scrape_metadata` out of unit databags, which contains the address